File: src/glare_ffi_queue.erl
-module(glare_ffi_queue).
-export([queue_send/2, queue_send_batch/2, queue_ack/1, queue_retry/1,
         queue_message_id/1, queue_message_timestamp/1, queue_message_body/1,
         queue_message_attempts/1]).

queue_send(_Queue, _Message) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_send_batch(_Queue, _Messages) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_ack(_Message) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_retry(_Message) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_message_id(_Message) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_message_timestamp(_Message) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_message_body(_Message) ->
    error({glare_error, "Not supported on Erlang target"}).

queue_message_attempts(_Message) ->
    error({glare_error, "Not supported on Erlang target"}).

File: src/glare_ffi_queue.mjs
import { Ok, Error } from "./gleam.mjs";

export async function queue_send(queue, message) {
  try {
    await queue.send(message);
    return new Ok(undefined);
  } catch (error) {
    return new Error(`${error}`);
  }
}

export async function queue_send_batch(queue, messages) {
  try {
    await queue.sendBatch(messages);
    return new Ok(undefined);
  } catch (error) {
    return new Error(`${error}`);
  }
}

export async function queue_ack(message) {
  try {
    await message.ack();
    return new Ok(undefined);
  } catch (error) {
    return new Error(`${error}`);
  }
}

export async function queue_retry(message) {
  try {
    await message.retry();
    return new Ok(undefined);
  } catch (error) {
    return new Error(`${error}`);
  }
}

export function queue_message_id(message) {
  return message.id;
}

export function queue_message_timestamp(message) {
  return message.timestamp;
}

export function queue_message_body(message) {
  return message.body;
}

export function queue_message_attempts(message) {
  return message.attempts;
}
